### ISSUE TITLE
Text processing bugs

### ISSUE THREAD

**Issue by Gillibald** [AvaloniaUI/Avalonia#8070]
This issue helps track bugs related to recent text processing changes. Feel free to post your findings here.

Successor of #7502

**Comment by Gillibald** [comment]
Posted by @danipen
@Gillibald, I found a missing feature. 

- [ ] For all textboxes implementation I checked, **double-clicking and dragging over text, allows you to extend the selection by word.**

### Example
https://user-images.githubusercontent.com/501613/159515582-c0a4f7e0-6e88-4799-a469-2a93a121fb4d.mov

### Avalonia's
https://user-images.githubusercontent.com/501613/159515549-88039fd5-3085-4427-8d2f-f200fa570aad.mov

**Comment by Gillibald** [comment]
- [ ] IME position update doesn't take scrolling into account
    https://user-images.githubusercontent.com/14960345/155470980-ad92f46e-b104-4b73-829b-ffe31f70c3f4.mp4

**Comment by adirh3** [comment]
@Gillibald #8053 works really well, checked with 250% DPI.
Only issue I am seeing now is TextBox caret is trying to run away 😆 
![image](https://user-images.githubusercontent.com/27368554/166492213-f3f11033-b7c2-429c-bb57-9722cc341366.png)
Thanks!

**Comment by Gillibald** [comment]
Probably some side effects I introduced with my hack-fix for the scaling issue

**Comment by danielmayost** [comment]
- [ ] [RTL] Input: `אאא AAA`:
![test](https://user-images.githubusercontent.com/41772276/166508794-76a314bf-534e-489f-b638-66c1f6d49ed3.gif)
:

**Comment by hez2010** [comment]
> - [ ] IME position update doesn't take scrolling into account
>     https://user-images.githubusercontent.com/14960345/155470980-ad92f46e-b104-4b73-829b-ffe31f70c3f4.mp4

This is okay to keep it unfixed. I found that even WinUI doesn't handle this case.

**Comment by danielmayost** [comment]
- [ ] text in `Flyouts` Tab exits outside the right border if is RTL FlowDirection:
<img src="https://user-images.githubusercontent.com/41772276/159263593-f20d4efd-7f50-40e0-9b6d-7bb5b52bf947.png" height=180 width=220/>

**Note:** This issue was previously fixed by  #7946 and is now back again after #8053.

**Comment by adirh3** [comment]
Found repro to the `TextBox` issue after #8053 4th/5th commit -
Add this in the `TextBoxPage.xaml` in the ControlCatalog
```
       <TextBox Width="200"
                 HorizontalAlignment="Stretch"
                 VerticalAlignment="Stretch"
                 VerticalContentAlignment="Center"
                 BorderBrush="Transparent"
                 Margin="0"
                 Padding="0"
                 Watermark="Watermark"
                 Text="Borderless text box" BorderThickness="0" CornerRadius="0" Background="Transparent"
                 TextAlignment="Left" >
          <TextBox.Resources>
            <Thickness x:Key="TextControlBorderThemeThicknessFocused">0</Thickness>
          </TextBox.Resources>
        </TextBox>
```
The issue is due to `VerticalContentAlignment="Center"`, I removed the borders to make it more noticeable, it's not related.
Empty the text, care going out of bounds - 
<img width="163" alt="image" src="https://user-images.githubusercontent.com/27368554/166819238-91361010-36b1-4af9-b833-610286ab9f4e.png">

**Comment by Gillibald** [comment]
> * [ ]  text in `Flyouts` Tab exits outside the right border if is RTL FlowDirection:
> 
> <img alt="" width="220/" height="180" src="https://user-images.githubusercontent.com/41772276/159263593-f20d4efd-7f50-40e0-9b6d-7bb5b52bf947.png">
> 
> **Note:** This issue was previously fixed by #7946 and is now back again after #8053.

Most likely padding isn't handled in the ArrangeOverride

**Comment by Gillibald** [comment]
![RTLSelection](https://user-images.githubusercontent.com/5928797/166939654-69a8fd32-c63f-4ecf-b05a-4b7ee30f37b4.gif)

**Comment by adirh3** [comment]
Seems like there are many issues with Chinese text after #8053 -

Missing word - 
![image](https://user-images.githubusercontent.com/27368554/167110692-29fb7e4e-ddf0-4688-b220-175c06e6c16f.png)

Empty columns in DataGrid - 
<img width="1001" alt="image" src="https://user-images.githubusercontent.com/27368554/167110863-291cbda9-358b-4801-b5ff-a7a6b0c12902.png">

Text wrapping cropped - 
<img width="844" alt="image" src="https://user-images.githubusercontent.com/27368554/167110956-05fd75a1-181e-44ab-bce5-57544e1a251e.png">

**Comment by Gillibald** [comment]
Could you provide me with some sample texts so I can add some unit tests? Just open up DevTools and copy property values like Width, Height, FontSize, TextWrapping, Text etc.

This would help a lot.

Will provide a new PR shortly

**Comment by adirh3** [comment]
Here is one example, I will add more later.

Text `在 Powershell 中运行`
FontSize	`20.16`
Font Family `Segoe UI`
FontWeight	`DemiBold`
HorizontalAlignment	`Center`
VerticalAlignment	`Stretch`
Margin	`4,4,4,8`
ClipToBounds	`True`
TextWrapping	`Wrap`
MaxLines `3`
TransformedBounds	```Bounds: 0, 0, 189.77777777777777, 26.666666666666668 Clip: 476.8888888888889, 200.44444444444443, 189.77777777777783, 26.666666666666657 Transform { {M11:1 M12:0} {M21:0 M22:1} {M31:476.8888888888889 M32:200.44444444444443} }	TransformedBounds? {TransformedBounds}```
DesiredSize	`197.77777777777777, 38.66666666666667`
Bounds	`96.88888888888889, 98.66666666666667, 189.77777777777777, 26.666666666666668`

**Comment by Gillibald** [comment]
I thought about adding some export (copy to clipboard) function to DevTools so providing this info will be easier.

**Comment by adirh3** [comment]
> I thought about adding some export (copy to clipboard) function to DevTools so providing this info will be easier.

That would be nice, you might also want (or already have) something to change scaling regardless of monitor/OS.
Will send next batch of examples after using #8094.

**Comment by maxkatz6** [comment]
I can reproduce https://github.com/AvaloniaUI/Avalonia/issues/8070#issuecomment-1116277384 issue on ControlCatalog CalendarDatePicker page:
![image](https://user-images.githubusercontent.com/3163374/169205031-b5c0a55f-eed1-4cf2-9b99-f368817240be.png)

1. 125% scaling
2. Vertically centered textbox
3. Empty text presenter

**Comment by adirh3** [comment]
@Gillibald sorry for not bringing more examples for a while, didn't have much time.
One important thing I noticed with latest master, is that there is a big memory leak of BidiAlgorithm instances -
![WhatsApp Image 2022-05-29 at 16 58 20](https://user-images.githubusercontent.com/27368554/170873154-fad86374-a9e8-4018-9479-b7a7d90057f7.jpg)
Not sure if it's already fixed in your new branch, haven't checked yet

**Comment by Gillibald** [comment]
Buffers are currently reused so we might clean up buffers at some point

**Comment by Gillibald** [comment]
If you are building from source you can alter this line: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/TextFormatting/TextFormatterImpl.cs#L180

Construct a new instance on every FormatLine call instead of reusing the ThreadLocal instance. That should free up memory but produces more GC pressure.

**Comment by adirh3** [comment]
> If you are building from source you can alter this line: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Media/TextFormatting/TextFormatterImpl.cs#L180
> 
> Construct a new instance on every FormatLine call instead of reusing the ThreadLocal instance. That should free up memory but produces more GC pressure.

Thanks! can confirm it worked.
You might want to change that in master, since it causes a big memory leak (seen up to 4GB in my app).

**Comment by Gillibald** [comment]
Thanks for testing. Will make the change.

**Comment by adirh3** [comment]
@Gillibald thanks, can confirm that 90% of the Chinese issues are fixed now.

Only once, very weird instance that there is an issue now -
Actual -
![image](https://user-images.githubusercontent.com/27368554/171236302-5edb9a94-52d1-4450-84e0-ec637656e5c4.png)
Expected -
![image](https://user-images.githubusercontent.com/27368554/171236363-8c88431c-4e22-45a1-b4e4-3eb0a1a67708.png)

Dump from the DevTools -

	Background	(null)	IBrush	Unset
	Bounds	0, 20, 602.2222222222222, 18.22222222222222	Rect	Direct
	Clip	(null)	Geometry	Unset
	ClipToBounds	True	Boolean	LocalValue
	Clock	(null)	IClock	Unset
	ContextFlyout	(null)	FlyoutBase	Unset
	ContextMenu	(null)	ContextMenu	Unset
	Cursor	(null)	Cursor	Unset
	DataContext	Blast.Core.Objects.Setting	Object {Setting}	Unset
	DesiredSize	602.2222222222222, 18.22222222222222	Size	Direct
	FocusAdorner	Avalonia.Markup.Xaml.Templates.FocusAdornerTemplate	ITemplate<IControl> {FocusAdornerTemplate}	Style
	Focusable	False	Boolean	Unset
	HasMirrorTransform	False	Boolean	Direct
	Height	NaN	Double	Unset
	HorizontalAlignment	Left	HorizontalAlignment	LocalValue
	Inlines	Avalonia.Controls.Documents.InlineCollection	InlineCollection	Direct
	IsEffectivelyEnabled	True	Boolean	Direct
	IsEnabled	True	Boolean	Unset
	IsFocused	False	Boolean	Direct
	IsHitTestVisible	True	Boolean	Unset
	IsKeyboardFocusWithin	False	Boolean	Direct
	IsPointerOver	False	Boolean	Direct
	IsVisible	True	Boolean	LocalValue
	Margin	0,0,0,0	Thickness	Unset
	MaxHeight	∞	Double	Unset
	MaxWidth	∞	Double	Unset
	MinHeight	0	Double	Unset
	MinWidth	0	Double	Unset
	Name	(null)	String	Direct
	Opacity	0.5	Double	LocalValue
	OpacityMask	(null)	IBrush	Unset
	Padding	0,0,0,0	Thickness	Unset
	Parent	Avalonia.Controls.DockPanel	IStyledElement {DockPanel}	Direct
	RenderTransform	(null)	ITransform	Unset
	RenderTransformOrigin	50%, 50%	RelativePoint	Unset
	Tag	(null)	Object	Unset
	TemplatedParent	(null)	ITemplatedControl	Direct
	Text	不对该路径进行索引和搜索。对于多种文件，可以用通配符和英文逗号隔开，例如：*.json,*file.tmp	String	Direct
	TextDecorations	(null)	TextDecorationCollection	Unset
	TransformedBounds	Bounds: 0, 0, 602.2222222222222, 18.22222222222222 Clip: 400.88888888888886, 326.6666666666667, 602.2222222222222, 18.22222222222223 Transform { {M11:1 M12:0} {M21:0 M22:1} {M31:400.88888888888886 M32:326.6666666666667} }	TransformedBounds? {TransformedBounds}	Direct
	Transitions	(null)	Transitions	Unset
	UseLayoutRounding	True	Boolean	Unset
	VerticalAlignment	Bottom	VerticalAlignment	LocalValue
	VisualParent	Avalonia.Controls.DockPanel	IVisual {DockPanel}	Direct
	Width	NaN	Double	Unset
	ZIndex	0	Int32	Unset
	__AvaloniaReserved::Classes::no-content-pointerover	False	Boolean	Unset
	__AvaloniaReserved::Classes::with-constraint	False	Boolean	Unset
	[AccessText.ShowAccessKey]	False	Boolean	Unset
	[AdornerLayer.AdornedElementInfo]	(null)	AdornedElementInfo	Unset
	[AdornerLayer.AdornedElement]	(null)	Visual	Unset
	[AdornerLayer.IsClipEnabled]	True	Boolean	Unset
	[AutomationProperties.AcceleratorKey]	(null)	String	Unset
	[AutomationProperties.AccessKey]	(null)	String	Unset
	[AutomationProperties.AccessibilityView]	Content	AccessibilityView	Unset
	[AutomationProperties.AutomationId]	(null)	String	Unset
	[AutomationProperties.ControlTypeOverride]	(null)	AutomationControlType?	Unset
	[AutomationProperties.HelpText]	(null)	String	Unset
	[AutomationProperties.IsColumnHeader]	False	Boolean	Unset
	[AutomationProperties.IsOffscreenBehavior]	Default	IsOffscreenBehavior	Unset
	[AutomationProperties.IsRequiredForForm]	False	Boolean	Unset
	[AutomationProperties.IsRowHeader]	False	Boolean	Unset
	[AutomationProperties.ItemStatus]	(null)	String	Unset
	[AutomationProperties.ItemType]	(null)	String	Unset
	[AutomationProperties.LabeledBy]	(null)	IControl	Unset
	[AutomationProperties.LiveSetting]	Off	AutomationLiveSetting	Unset
	[AutomationProperties.Name]	(null)	String	Unset
	[AutomationProperties.PositionInSet]	-1	Int32	Unset
	[AutomationProperties.SizeOfSet]	-1	Int32	Unset
	[Canvas.Bottom]	NaN	Double	Unset
	[Canvas.Left]	NaN	Double	Unset
	[Canvas.Right]	NaN	Double	Unset
	[Canvas.Top]	NaN	Double	Unset
	[Control.FlowDirection]	LeftToRight	FlowDirection	Unset
	[DataGridFrozenGrid.IsFrozen]	False	Boolean	Unset
	[DataValidationErrors.Errors]	(null)	IEnumerable<Object>	Unset
	[DataValidationErrors.HasErrors]	False	Boolean	Unset
	[DefinitionBase.PrivateSharedSizeScope]	(null)	SharedSizeScope	Unset
	[DefinitionBase.SharedSizeGroup]	(null)	String	Unset
	[Design.DataContext]	(null)	Object	Unset
	[Design.DesignStyle]	(null)	IStyle	Unset
	[Design.Height]	0	Double	Unset
	[Design.PreviewWith]	(null)	Control	Unset
	[Design.Width]	0	Double	Unset
	[DockPanel.Dock]	Top	Dock	LocalValue
	[DragDrop.AllowDrop]	False	Boolean	Unset
	[FlyoutBase.AttachedFlyout]	(null)	FlyoutBase	Unset
	[Grid.ColumnSpan]	1	Int32	Unset
	[Grid.Column]	0	Int32	Unset
	[Grid.IsSharedSizeScope]	False	Boolean	Unset
	[Grid.RowSpan]	1	Int32	Unset
	[Grid.Row]	0	Int32	Unset
	[HotKeyManager.HotKey]	(null)	KeyGesture	Unset
	[InputMethod.IsInputMethodEnabled]	True	Boolean	Unset
	[KeyboardNavigation.IsTabStop]	True	Boolean	Unset
	[KeyboardNavigation.TabIndex]	2147483647	Int32	Unset
	[KeyboardNavigation.TabNavigation]	Continue	KeyboardNavigationMode	Unset
	[KeyboardNavigation.TabOnceActiveElement]	(null)	IInputElement	Unset
	[NameScope.NameScope]	(null)	INameScope	Unset
	[RenderOptions.BitmapInterpolationMode]	MediumQuality	BitmapInterpolationMode	Unset
	[ScrollViewer.AllowAutoHide]	True	Boolean	Unset
	[ScrollViewer.HorizontalScrollBarVisibility]	Disabled	ScrollBarVisibility	Unset
	[ScrollViewer.IsScrollChainingEnabled]	True	Boolean	Unset
	[ScrollViewer.VerticalScrollBarVisibility]	Auto	ScrollBarVisibility	Unset
	[TemplatedControl.IsTemplateFocusTarget]	False	Boolean	Unset
	[TextBlock.BaselineOffset]	0	Double	Unset
	[TextBlock.LineHeight]	NaN	Double	Unset
	[TextBlock.MaxLines]	0	Int32	LocalValue
	[TextBlock.TextAlignment]	Left	TextAlignment	Unset
	[TextBlock.TextTrimming]	None	TextTrimming {TextNoneTrimming}	LocalValue
	[TextBlock.TextWrapping]	Wrap	TextWrapping	LocalValue
	[TextElement.FontFamily]	(unset)	FontFamily {UnsetValueType}	Unset
	[TextElement.FontSize]	13.86	Double	LocalValue
	[TextElement.FontStretch]	Normal	FontStretch	Unset
	[TextElement.FontStyle]	Normal	FontStyle	Unset
	[TextElement.FontWeight]	(unset)	FontWeight {UnsetValueType}	Unset
	[TextElement.Foreground]	White	IBrush {SolidColorBrush}	Unset
	[TextInputOptions.AutoCapitalization]	False	Boolean	Unset
	[TextInputOptions.ContentType]	Normal	TextInputContentType	Unset
	[TextInputOptions.IsSensitive]	False	Boolean	Unset
	[TextInputOptions.Lowercase]	False	Boolean	Unset
	[TextInputOptions.Multiline]	False	Boolean	Unset
	[TextInputOptions.Uppercase]	False	Boolean	Unset
	[ToolTip.HorizontalOffset]	0	Double	Unset
	[ToolTip.IsOpen]	False	Boolean	Unset
	[ToolTip.Placement]	Pointer	PlacementMode	Unset
	[ToolTip.ShowDelay]	400	Int32	Unset
	[ToolTip.Tip]	(null)	Object	Unset
	[ToolTip.ToolTip]	(null)	ToolTip	Unset
	[ToolTip.VerticalOffset]	20	Double	Unset
	Background	(null)	IBrush	
	BaselineOffset	0	Double	
	Bounds	0, 20, 602.2222222222222, 18.22222222222222	Rect	
	Classes	Avalonia.Controls.Classes	Classes	
	Clip	(null)	Geometry	
	ClipToBounds	True	Boolean	
	Clock	(null)	IClock	
	ContextFlyout	(null)	FlyoutBase	
	ContextMenu	(null)	ContextMenu	
	Cursor	(null)	Cursor	
	DataContext	Blast.Core.Objects.Setting	Object {Setting}	
	DataTemplates	Avalonia.Controls.Templates.DataTemplates	DataTemplates	
	DesiredSize	602.2222222222222, 18.22222222222222	Size	
	FlowDirection	LeftToRight	FlowDirection	
	FocusAdorner	Avalonia.Markup.Xaml.Templates.FocusAdornerTemplate	ITemplate<IControl> {FocusAdornerTemplate}	
	Focusable	False	Boolean	
	FontFamily	Segoe UI Variable Text Light	FontFamily	
	FontSize	13.86	Double	
	FontStretch	Normal	FontStretch	
	FontStyle	Normal	FontStyle	
	FontWeight	Normal	FontWeight	
	Foreground	White	IBrush {SolidColorBrush}	
	GestureRecognizers	Avalonia.Input.GestureRecognizers.GestureRecognizerCollection	GestureRecognizerCollection	
	HasMirrorTransform	False	Boolean	
	Height	NaN	Double	
	HorizontalAlignment	Left	HorizontalAlignment	
	IControl.Parent	Avalonia.Controls.DockPanel	IControl {DockPanel}	
	IDataContextProvider.DataContext	Blast.Core.Objects.Setting	Object {Setting}	
	IDataTemplateHost.DataTemplates	Avalonia.Controls.Templates.DataTemplates	DataTemplates	
	IDataTemplateHost.IsDataTemplatesInitialized	True	Boolean	
	IInputElement.Cursor	(null)	Cursor	
	IInputElement.Focusable	False	Boolean	
	IInputElement.IsEffectivelyEnabled	True	Boolean	
	IInputElement.IsEnabled	True	Boolean	
	IInputElement.IsFocused	False	Boolean	
	IInputElement.IsHitTestVisible	True	Boolean	
	IInputElement.IsKeyboardFocusWithin	False	Boolean	
	IInputElement.IsPointerOver	False	Boolean	
	IInputElement.KeyBindings	System.Collections.Generic.List`1[Avalonia.Input.KeyBinding]	List<KeyBinding>	
	IInteractive.InteractiveParent	Avalonia.Controls.DockPanel	IInteractive {DockPanel}	
	ILayoutable.DesiredSize	602.2222222222222, 18.22222222222222	Size	
	ILayoutable.Height	NaN	Double	
	ILayoutable.HorizontalAlignment	Left	HorizontalAlignment	
	ILayoutable.IsArrangeValid	True	Boolean	
	ILayoutable.IsMeasureValid	True	Boolean	
	ILayoutable.Margin	0,0,0,0	Thickness	
	ILayoutable.MaxHeight	∞	Double	
	ILayoutable.MaxWidth	∞	Double	
	ILayoutable.MinHeight	0	Double	
	ILayoutable.MinWidth	0	Double	
	ILayoutable.PreviousArrange	0, 20, 603.5555555555555, 18.22222222222222	Rect? {Rect}	
	ILayoutable.PreviousMeasure	602.6266666666668, Infinity	Size? {Size}	
	ILayoutable.VerticalAlignment	Bottom	VerticalAlignment	
	ILayoutable.Width	NaN	Double	
	ILogical.IsAttachedToLogicalTree	True	Boolean	
	ILogical.LogicalChildren	Avalonia.Collections.AvaloniaList`1[Avalonia.LogicalTree.ILogical]	IAvaloniaReadOnlyList<ILogical> {AvaloniaList<ILogical>}	
	ILogical.LogicalParent	Avalonia.Controls.DockPanel	ILogical {DockPanel}	
	INamed.Name	(null)	String	
	IResourceNode.HasResources	False	Boolean	
	IStyleHost.IsStylesInitialized	True	Boolean	
	IStyleHost.Styles	Avalonia.Styling.Styles	Styles	
	IStyleHost.StylingParent	Avalonia.Controls.DockPanel	IStyleHost {DockPanel}	
	IStyleable.Classes	Avalonia.Controls.Classes	IAvaloniaReadOnlyList<String> {Classes}	
	IStyleable.StyleKey	Avalonia.Controls.TextBlock	Type {RuntimeType}	
	IStyleable.TemplatedParent	(null)	ITemplatedControl	
	IStyledElement.Classes	Avalonia.Controls.Classes	Classes	
	IStyledElement.IsInitialized	True	Boolean	
	IStyledElement.Parent	Avalonia.Controls.DockPanel	IStyledElement {DockPanel}	
	IVisual.Bounds	0, 20, 602.2222222222222, 18.22222222222222	Rect	
	IVisual.Clip	(null)	Geometry	
	IVisual.ClipToBounds	True	Boolean	
	IVisual.HasMirrorTransform	False	Boolean	
	IVisual.IsAttachedToVisualTree	True	Boolean	
	IVisual.IsEffectivelyVisible	True	Boolean	
	IVisual.IsVisible	True	Boolean	
	IVisual.Opacity	0.5	Double	
	IVisual.OpacityMask	(null)	IBrush	
	IVisual.RenderTransform	(null)	ITransform	
	IVisual.RenderTransformOrigin	50%, 50%	RelativePoint	
	IVisual.TransformedBounds	Bounds: 0, 0, 602.2222222222222, 18.22222222222222 Clip: 400.88888888888886, 326.6666666666667, 602.2222222222222, 18.22222222222223 Transform { {M11:1 M12:0} {M21:0 M22:1} {M31:400.88888888888886 M32:326.6666666666667} }	TransformedBounds? {TransformedBounds}	
	IVisual.VisualChildren	Avalonia.Collections.AvaloniaList`1[Avalonia.VisualTree.IVisual]	IAvaloniaReadOnlyList<IVisual> {AvaloniaList<IVisual>}	
	IVisual.VisualParent	Avalonia.Controls.DockPanel	IVisual {DockPanel}	
	IVisual.VisualRoot	Blast.Views.SettingsWindow	IRenderRoot {SettingsWindow}	
	IVisual.ZIndex	0	Int32	
	Inlines	Avalonia.Controls.Documents.InlineCollection	InlineCollection	
	IsArrangeValid	True	Boolean	
	IsEffectivelyEnabled	True	Boolean	
	IsEffectivelyVisible	True	Boolean	
	IsEnabled	True	Boolean	
	IsFocused	False	Boolean	
	IsHitTestVisible	True	Boolean	
	IsInitialized	True	Boolean	
	IsKeyboardFocusWithin	False	Boolean	
	IsMeasureValid	True	Boolean	
	IsPointerOver	False	Boolean	
	IsTabStop	True	Boolean	
	IsVisible	True	Boolean	
	KeyBindings	System.Collections.Generic.List`1[Avalonia.Input.KeyBinding]	List<KeyBinding>	
	LineHeight	NaN	Double	
	Margin	0,0,0,0	Thickness	
	MaxHeight	∞	Double	
	MaxLines	0	Int32	
	MaxWidth	∞	Double	
	MinHeight	0	Double	
	MinWidth	0	Double	
	Name	(null)	String	
	Opacity	0.5	Double	
	OpacityMask	(null)	IBrush	
	Padding	0,0,0,0	Thickness	
	Parent	Avalonia.Controls.DockPanel	IControl {DockPanel}	
	Parent	Avalonia.Controls.DockPanel	IStyledElement {DockPanel}	
	RenderTransform	(null)	ITransform	
	RenderTransformOrigin	50%, 50%	RelativePoint	
	Resources	Avalonia.Controls.ResourceDictionary	IResourceDictionary {ResourceDictionary}	
	Styles	Avalonia.Styling.Styles	Styles	
	TabIndex	2147483647	Int32	
	Tag	(null)	Object	
	TemplatedParent	(null)	ITemplatedControl	
	Text	不对该路径进行索引和搜索。对于多种文件，可以用通配符和英文逗号隔开，例如：*.json,*file.tmp	String	
	TextAlignment	Left	TextAlignment	
	TextDecorations	(null)	TextDecorationCollection	
	TextLayout	Avalonia.Media.TextFormatting.TextLayout	TextLayout	
	TextTrimming	None	TextTrimming {TextNoneTrimming}	
	TextWrapping	Wrap	TextWrapping	
	TransformedBounds	Bounds: 0, 0, 602.2222222222222, 18.22222222222222 Clip: 400.88888888888886, 326.6666666666667, 602.2222222222222, 18.22222222222223 Transform { {M11:1 M12:0} {M21:0 M22:1} {M31:400.88888888888886 M32:326.6666666666667} }	TransformedBounds? {TransformedBounds}	
	Transitions	(null)	Transitions	
	UseLayoutRounding	True	Boolean	
	VerticalAlignment	Bottom	VerticalAlignment	
	Width	NaN	Double	
	ZIndex	0	Int32	

Let me know if more info is needed.

**Comment by danielmayost** [comment]
- [ ] [RTL] Since the last merge, text selection has gone completely wrong.

**Comment by Gillibald** [comment]
Text selection on TextLine level got completely reworked for WPF compatibility. We need to add more unit tests in that area.

**Comment by adirh3** [comment]
@Gillibald, just wondering, does #8347 address the Chinese text issues?

**Comment by Gillibald** [comment]
What issues are you referring to? That PR just introduced some limited text justification algorithm and introduced more text alignment options.

**Comment by Gillibald** [comment]
#8370 fixes another scaling issue. Will finish this PR shortly.

**Comment by danielmayost** [comment]
@Gillibald There are other problems some of which have been marked here as solved, I will list them again?

**Comment by Gillibald** [comment]
Y feel free to list them again. RTL text selection still needs more test coverage because it got a major rework.

**Comment by Gillibald** [comment]
This isn't properly selectable `<Run>أَبْجَدِيَّة عَرَبِيَّة</Run>`

**Comment by danielmayost** [comment]
- [ ] After Unicode 14 I think, the brackets and delimiting characters of sorts, are reversed in RTL:
![image](https://user-images.githubusercontent.com/41772276/181090465-746812c5-0942-44d3-aa3d-1841121a6171.png)
- [ ] `TextAlignment="Justify"` in RTL mode don't work as expected, the last line aligned from left to right instead from right to left:
![image](https://user-images.githubusercontent.com/41772276/181091962-5c838dc7-5bd7-4445-9b60-f06db281c8da.png)

There is more, I will update later.

**Comment by maxkatz6** [comment]
@Gillibald not sure if it's a bug or expected behavior, but string like `C:\привет мир\幸福\幸福` is displayed differently in the TextBox:
![image](https://user-images.githubusercontent.com/3163374/181581602-d7a3ad8b-c4d1-4b52-8b1b-39461945434c.png)

"\\" is replaced with 
![image](https://user-images.githubusercontent.com/3163374/181582033-abb09d4e-526d-456c-aa67-9bcf8365d8c3.png)

**Comment by adirh3** [comment]
@Gillibald `Inlines` binding suppose to work? Invoking property change on `Inlines` does nothing (Text not changing), but if I bind the `Text` in addition + invoke property change it's working.
Also seems like after recent changes `HasComplexContent` throws NRE because `Inlines` is not initialized yet (probably related to first issue)

To repro simply do -

XAML
```
  <DockPanel DockPanel.Dock="Top" LastChildFill="False">
            <RichTextBlock Name="RichTextBlock"
                           TextAlignment="Left"
                           Margin="4,4,4,8"
                           ClipToBounds="True"
                           FontSize="30"
                           MaxLines="3"
                           TextWrapping="Wrap"
                           FontWeight="DemiBold"
                           Text="{Binding Text}"
                           Inlines="{Binding InlineCollection, Mode=OneWay}">
            </RichTextBlock>
            <Button Click="Button_OnClick">Update inlines</Button>
            <Button Click="TextButton_OnClick">Update text</Button>
        </DockPanel>
        ```
        
        C#
 ```
       public class ViewModel : INotifyPropertyChanged
    {
        private InlineCollection _inlineCollection;
        private string _text;

        public string Text
        {
            get => _text;
            set
            {
                if (value == _text) return;
                _text = value;
                OnPropertyChanged();
            }
        }

        public InlineCollection InlineCollection
        {
            get => _inlineCollection;
            set
            {
                if (Equals(value, _inlineCollection)) return;
                _inlineCollection = value;
                OnPropertyChanged();
            }
        }

        public event PropertyChangedEventHandler? PropertyChanged;

        [NotifyPropertyChangedInvocator]
        protected virtual void OnPropertyChanged([CallerMemberName] string? propertyName = null)
        {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
        }
    }

    public partial class MainWindow : Window
    {
        private readonly ViewModel _dc;

        public MainWindow()
        {
            InitializeComponent();

            DataContext = _dc = new ViewModel
            {
                Text = "lala",
                InlineCollection = new InlineCollection
                {
                    new Run("yyy"),
                    new Run("test2") {FontWeight = FontWeight.Bold},
                }
            };
        }

        private void Button_OnClick(object? sender, RoutedEventArgs e)
        {
            _dc.InlineCollection = new InlineCollection
            {
                new Run(""),
                new Run("test3") {FontWeight = FontWeight.Bold},
            };
            // _dc.Text = "nununu";
        }

        private void TextButton_OnClick(object? sender, RoutedEventArgs e)
        {
            _dc.Text = "nununu";
        }
    }
    ```

**Comment by Gillibald** [comment]
I guess bindings still produce null values so I have to make Inlines nullable and properly handle that case

**Comment by Gillibald** [comment]
This line is wrong: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/RichTextBlock.cs#L540

Outch 😅

**Comment by timunie** [comment]
> I guess bindings still produce null values so I have to make Inlines nullable and properly handle that case.
> 
> Setting Text and assigning Inlines isn't supported. How should we deal with that? Should the text be added to the InlineCollection? At the start or end?

Tbh I think this should throw a compile time error. Like in WPF setting Items and ItemsSource. Your text may be added to inlines collection if you have it anyway in your VM?!

**Comment by adirh3** [comment]
> I guess bindings still produce null values so I have to make Inlines nullable and properly handle that case.
> 
> Setting Text and assigning Inlines isn't supported. How should we deal with that? Should the text be added to the InlineCollection? At the start or end?

The only reason I set `Text` is because `Inlines` wasn't updating, but I see now you already identified the cause :)

**Comment by adirh3** [comment]
@Gillibald When `TextTrimming` is not set, it causes `NotImplementedException`, if using `RichTextBlock` with `MaxLines` when there are too many characters.
Seems to be due to this line -
https://github.com/AvaloniaUI/Avalonia/blob/ad6eef6dd41d8d454b89053fd442bfbb46d694f5/src/Avalonia.Base/Media/TextFormatting/TextLayout.cs#L484

**Comment by Gillibald** [comment]
We need to cleanup all children when the old collection is detached from the InlinesHost

**Comment by Gillibald** [comment]
This passes:

```
var target = new RichTextBlock();

var run = new Run("Hello");

target.Inlines.Add(run);

target.Measure(Size.Infinity);

Assert.True(target.IsMeasureValid);

target.Inlines = null;

Assert.Null(run.Parent);

target.Inlines = new InlineCollection { run };

Assert.Equal(target, run.Parent);
```

**Comment by Gillibald** [comment]
#8626
![UpdateInlines](https://user-images.githubusercontent.com/5928797/181736318-41980bf3-4472-4b13-9814-f2a2f99f8697.gif)

**Comment by danielmayost** [comment]
- [ ] `<TextBox AcceptsReturn="True" TextWrapping="Wrap"/>in ` RTL mode enters an infinite loop when typing `א` and then `Enter`.

**Comment by Gillibald** [comment]
#8671

**Comment by adirh3** [comment]
@Gillibald the parent issue in `InlinesCollection` still exists, FYI it seems to be solved by changing to this -
```
      private void OnParentChanged(ILogical? parent)
        {
            foreach (var child in this)
            {
                if (parent != null)
                    ((ISetLogicalParent)child).SetParent(null);
                ((ISetLogicalParent)child).SetParent(parent);
            }
        }
```

**Comment by Gillibald** [comment]
@adirh3 I am not sure this is really needed because inlines should be cleaned up when they are removed from the InlineCollection and when the collection itself is reparented. Are you sure that the inlines aren't used anymore?

**Comment by adirh3** [comment]
> @adirh3 I am not sure this is really needed because inlines should be cleaned up when they are removed from the InlineCollection and when the collection itself is reparented. Are you sure that the inlines aren't used anymore?

Unless I add that it will crash. I am creating the InlineCollection from code and put it my vm, so maybe I am somehow using it wrong?
Will see if I can make a simple rerpro and post here later

**Comment by Gillibald** [comment]
Assigning inlines should cleanup the old collection if not that is a bug. So in the end each child in the old collection should not have a parent anymore. Will add a test to verify.

**Comment by jl0pd** [comment]
Deleting text with backspace results in square sign. When any key is pressed, input event is raised once, but when backspace is hit - event raised twice: first with backspace button click, second with `\b` text. Textbox is located inside WPF window

![avalonia_tb](https://user-images.githubusercontent.com/36221916/186331525-3d77abae-3bb9-4b1c-9d56-305b5095b0de.gif)

**Comment by ahopper** [comment]
using 11.0.999-cibuild0024336-beta TextBoxes bound to a double have some strange behaviour.
Sometimes the decimal point is ignored if you type quickly.
If you type '1.000003' and then delete the 3 it jumps to displaying '1'

**Comment by DarshZero** [comment]
the arabic is rendered except one character called mem "م"
![image](https://user-images.githubusercontent.com/31729048/198735403-e8d432e0-85e3-4e83-8f2a-49f81ea9a18b.png)

**Comment by Gillibald** [comment]
Are you sure this isn't an issue of the used font? Have you tried a different font? Feel free to post me the actual used text here so I can debug this.

**Comment by DarshZero** [comment]
i use the  default font but after changing the font, the text is rendered well without any   problems.

**Comment by adirh3** [comment]
@Gillibald using latest previews, text seems to not clip to bounds when using small font size (<10) -
![image](https://user-images.githubusercontent.com/27368554/199311173-f1deba81-bf88-433c-b36a-4515d04e97bd.png)

Will try to make a repro.

**Comment by AhmedZero** [comment]
![image](https://user-images.githubusercontent.com/26312527/202324514-a8af76df-a5a0-4729-8fe4-9b01d35bafab.png)
Reorder Bi-Directional, it should be
ليك 5% خصم اضافي مع Valu.
there are libraries that use that like [Fribidi ](https://github.com/fribidi/fribidi) and [crubit](https://github.com/google/crubit)

**Comment by AhmedZero** [comment]
![image](https://user-images.githubusercontent.com/26312527/202325240-1c25055e-607f-4808-8de4-b13f1065c9ad.png)

please support bi-directional reorder, it should be
![image](https://user-images.githubusercontent.com/26312527/202325345-75b631eb-8634-4d8c-8bc4-77c8a89e4b7c.png)

**Comment by danielmayost** [comment]
@AhmedZero do you set the `FlowDirection` to `RightToLeft`?

**Comment by Gillibald** [comment]
Without posting the actual text sequence I can't verify anything

**Comment by AhmedZero** [comment]
ليك 5% خصم اضافي مع valU

**Comment by danielmayost** [comment]
it's work well. just set the `FlowDirection` property to `RightToLeft`.

**Comment by AhmedZero** [comment]
I mean like GitHub did without specifying the direction.
GitHub writes it correctly except percent sign cause.

**Comment by AhmedZero** [comment]
I mean that because sometime the cell will be written in English and sometimes in Arabic.

**Comment by danielmayost** [comment]
Basically, i think there is no algorithm that deduces from the text, how it should appear. This is something you will have to provide. What some platforms do is automatic identification based on the first character, it's a gimmick that doesn't always give the result you want (for example a line starting with `A:` and then a text in Arabic.

But maybe it would be nice to add to Avalonia, not something of a high priority.

**Comment by hez2010** [comment]
The FlowDirection should be automatically detected for each line, not per TextBox.

For example:

ليك 5% خصم اضافي مع valU
A: ليك 5% خصم اضافي مع valU

**Comment by Gillibald** [comment]
That is possible y

**Comment by maxkatz6** [comment]
@Gillibald found another issue on the https://github.com/AvaloniaUI/Avalonia/pull/9526 branch. But I believe it should be reproducible on the master as well, as it's more about text rendering and not ime.
Problematic text:
```
호호호호호호
```
(on mac with 2-set korean input by pressing G and H keys repeatedly).

It works as expected if I copy paste this text into the text box with English text:
![Screen Shot 2023-01-23 at 9 51 35 PM](https://user-images.githubusercontent.com/3163374/214202716-b375bdb7-efc9-4a16-aa8c-d650482eca61.png)

But it gets squished together when I paste it into the text box with Chinese text: 
![Screen Shot 2023-01-23 at 9 52 27 PM](https://user-images.githubusercontent.com/3163374/214202794-09cfa1ea-6b22-4650-90a2-08c66aed08c7.png)

**Comment by Gillibald** [comment]
Thanks will have a look

**Comment by DarshZero** [comment]
![image](https://github.com/AvaloniaUI/Avalonia/assets/31729048/dfc77e7e-cc02-4e97-bb9e-02f4896f6777)
in Arabic language:
the first character is not fully rendered.

**Comment by timunie** [comment]
@DarshZero please try again with latest nightly. If issue persists, please open a new bug as closed issues are likely to be overseen. Thx 🙏

**Comment by Gillibald** [comment]
@DarshZero Please provide a repro that contains actual input sequence and the desired output
